### PR TITLE
fix: multiple bug fixes in Wave Generator

### LINKED ISSUE #1181: Wave generator not saving its state on back press
**Actual Behaviour**

Wave generator loses its values everytime it is reopened

**Expected Behaviour**

The wave generator common should not initialized only once in activity

**Steps to reproduce it**

Prevent reinitialization of Wave Generator Common

**Would you like to work on the issue?**

Yes

### LINKED ISSUE #1182: Digital wave generator pins not working
**Actual Behaviour**

Digital wave generating pins SQR1, SQR2, SQR3 and SQR4 aren't working properly i.e. SQR1 pin isn't able to set proper frequency and duty cycle whereas all other pins aren't working at all.

**Expected Behaviour**

All pins should function properly

**Steps to reproduce it**

Go to Wave Generator instrument -> Analyze

**LogCat for the issue**

Not applicable

**Screenshots of the issue**

screenshot_20180704-022603

**Would you like to work on the issue?**

@Avjeet Would you like to work on this issue?

### LINKED ISSUE #1183: Bug in sqr PWM signal
**Actual Behaviour**

The sqr PWM signal can only be produced if duty cycle of all the SQR pins is non zero positive number.

**Expected Behaviour**

If any duty cycle is zero for any SQR pin, its value should be set to some default non zero value but the pin should be disabled so that the signal can produce on rest of the SQR pins.

**Steps to reproduce it**

Make changes in Science Lab class and Wave generator for resolving PWM signal functionality.

**Would you like to work on the issue?**

Yes

### PR DESCRIPTION
**Pull request by Avjeet**
Fixes #1183 
Fixes #1181 

**Changes**:

- Resolved bug in Digital signal now producing properly (If the duty cycle is zero then that SQR Pin will 
   be activated else it will be deactivated)
- Wave Generator now not losing its saved values as WaveGenerator Common is not reinitialized 
   every time
- Changed naming of SQ pins to the SQR pins.
- Optimized some code and made it compact.

**Screenshot/s for the changes**: Not a UI change

**Checklist**: [Please tick following check boxes with `[x]` if the respective task is completed]
- [x] I have used resources from `strings.xml`, `dimens.xml` and `colors.xml` without hard-coding them
- [x] No modifications done at the end of resource files `strings.xml`, `dimens.xml` or `colors.xml`
- [x] I have reformatted code in every file included in this PR [<kbd>CTRL</kbd>+<kbd>ALT</kbd>+<kbd>L</kbd>]
- [x] My code does not contain any extra lines or extra spaces
- [x] I have requested reviews from other members

**APK for testing**:
[pwm_sqr_fix.zip](https://github.com/fossasia/pslab-android/files/2171575/pwm_sqr_fix.zip)

### CODE REVIEW

**Comment by Avjeet** [comment]
@harsh-2711 @Vikum94 @CloudyPadmal @abhinavraj23 Plz review it with your device and

**Comment by CloudyPadmal** [comment]
@harsh-2711 can you please check if this resolves the issues? Also @abhinavraj23 , @Vikum94 please test this with your devices

**Comment by Avjeet** [comment]
@harsh-2711 plz review this...asap

File: app/src/main/java/org/fossasia/pslab/communication/ScienceLab.java [hunk 1]
```diff
@@ -1075,7 +1075,7 @@ private int calcCHOSA(String channelName) {
     public double getVoltage(String channelName, Integer sample) {
         this.voltmeterAutoRange(channelName);
         double Voltage = this.getAverageVoltage(channelName, sample);
-        if (channelName.equals("CH2") || channelName.equals("CH1") ) {
+        if (channelName.equals("CH2") || channelName.equals("CH1")) {
```
Comment (Vikum94) [inline app/src/main/java/org/fossasia/pslab/communication/ScienceLab.java]: It's better to use string literal first. "CH2".equals(channelName)

Comment (Avjeet) [reply app/src/main/java/org/fossasia/pslab/communication/ScienceLab.java]: This specific change is due to refractor code.

**Comment by Avjeet** [comment]
@Vikum94 
- First connect the SQ1 and SQ2 pins to CH1 and CH2 .
- Select mode to PWM and then set the frequency.
- Set the duty cycle for both pins and click view to see if PWM signal is producing

**Comment by CloudyPadmal** [comment]
Looks OK. @Vikum94 @harsh-2711 @abhinavraj23 can you test this with your devices and confirm functionality?

**Comment by Avjeet** [comment]
@harsh-2711 @abhinavraj23 @Vikum94 @CloudyPadmal plz guys review it and merge it asap

**Comment by Avjeet** [comment]
@CloudyPadmal can you plz make sure this PR is merged before this week release if any bugs later I will resolve them but this PR is important for wave generator.
@harsh-2711 @Vikum94 @abhinavraj23 guys plz review this

**Comment by harsh-2711** [comment]
@Avjeet SQR1 and SQR2 pins aren't working with Logic Analyzer. Also, here is the output that I got after connecting them to CH1 and CH2 pins

screenshot_20180711-181655
screenshot_20180711-181632

So, I don't think this type of waves will work in digital section. Maybe you can use the code from `Control Unit` as it was completely working fine with `Logic Analyzer`

**Comment by Avjeet** [comment]
@harsh-2711 Its been eight days since I made this PR next time plz review a little faster anyways I look into it.

**Comment by Avjeet** [comment]
@harsh-2711 I have replaced the control activity with the wave generator and I am getting the same output as I m getting with the wave generator. 
@CloudyPadmal @abhinavraj23 @Vikum94 could u guys also plz also test this.

**Comment by harsh-2711** [comment]
@Avjeet This was the output when I used `Control Unit` for wave generation

screenshot_20180713-193851

I think you might be doing something wrong.

**Comment by Avjeet** [comment]
@harsh-2711 yeah this wave is produced with current wave generator

**Comment by harsh-2711** [comment]
@Avjeet Can you please remove reference 0f issue #1182 as this PR isn't solving that issue completely. I have worked on that issue and will soon make a PR for it.

**Comment by Avjeet** [comment]
@harsh-2711 done

**Comment by woshikie** [comment]
image
Above is my configuration for PWM on SQR1/2/3/4
image
Above is what I get when i connect SQR1/2/3/4 to CH1 

Looks weird, is it intended?

**Comment by woshikie** [comment]
This PR was referenced on #1078, does this fix that issue as well?

**Comment by CloudyPadmal** [comment]
@woshikie when CH1 is connected to SQR1/2/3/4 , it should display a square wave. @Avjeet are there any modifications done to that?

**Comment by woshikie** [comment]
@CloudyPadmal do you mean for PWM or Square Wave? It looks fine on Square Wave mode, but PWM mode looks odd.

**Comment by Avjeet** [comment]
@CloudyPadmal I used the same code from control advanced for sqr pwm mode so I think there is some issue with the communication module.
@woshikie I certainly resolve that unit issue

**Comment by Avjeet** [comment]
@woshikie the control unit advanced is converted into wave generator and when I checked with that it is also giving the same issue.That is why I think there is issue with communication class

**Comment by woshikie** [comment]
@Avjeet Can you please rephrase yourself? I am unable to understand fully what you are trying to convey.

**Comment by Avjeet** [comment]
The wave generator uses the same code as initially used by the Control advanced unit.
For testing why sqr PWM signal is producing like this I tried to use that control advanced unit and I m getting the same straight line in PWM signal as from current wave generator.
Therefore I think the issue is not with wave generator it is in the communication class or science lab class.

**Comment by woshikie** [comment]
Okay perhaps @CloudyPadmal could look into this issue with you?

**Comment by Avjeet** [comment]
@woshikie yeah that would be really helpful

**Comment by CloudyPadmal** [comment]
We can merge this now and proceed

**Comment by Avjeet** [comment]
@CloudyPadmal yeah it has some other bug fixed too

**Pull request merged**